src/backtrace.rs

```diff
@@ -0,0 +1,107 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use alloc::fmt::{Display, Formatter, Result};
+use core::arch::asm;
+use core::mem::size_of;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub enum BTReturnAddress {
+    ReturnAddress(u64),
+    InvalidFramePointer(u64),
+    InvalidReturnAddress(u64),
+}
+
+impl Display for BTReturnAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            BTReturnAddress::ReturnAddress(addr) => {
+                writeln!(f, "0x{addr:x}")
+            }
+            BTReturnAddress::InvalidFramePointer(addr) => {
+                writeln!(f, "Invalid frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::InvalidReturnAddress(addr) => {
+                writeln!(f, "Null address at frame pointer: 0x{addr:x}")
+            }
+        }
+    }
+}
+
+pub struct BackTrace {
+    fp: Option<u64>,
+    stack_start: u64,
+    stack_end: u64,
+    stack: &'static [u64],
+}
+
+impl BackTrace {
+    fn new(fp: u64) -> Self {
+        let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+        let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+        let stack = unsafe {
+            core::slice::from_raw_parts(
+                stack_start as *const u64,
+                (stack_end - stack_start) as usize,
+            )
+        };
+
+        Self {
+            fp: Some(fp),
+            stack_start,
+            stack_end,
+            stack,
+        }
+    }
+}
+
+impl Iterator for BackTrace {
+    type Item = BTReturnAddress;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(mut fp) = self.fp {
+            if fp < self.stack_start || fp > self.stack_end {
+                self.fp = None;
+                return Some(BTReturnAddress::InvalidFramePointer(fp));
+            }
+
+            let offset = ((fp - self.stack_start - 2) / (size_of::<u64>() as u64)) as usize;
+            if offset > 0 &&
```

Comment:
this is going to be true unless `fp == self.stack_start`. i think if you actually want to catch an `fp` below `stack_start` you need to do a `checked_sub()` above

Reply:
Yeah, seems like we will encounter this because we removed the check for ` fp - 2` that guarded against the underflow. Using something like the following should set offset to 0 without a panic, but we may want to check for the underflow and simply bail.

` let offset = ((fp.checked_sub(self.stack_start).unwrap_or(2) - 2) / (size_of::<u64>() as u64)) as usize;`